src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/Source.cs

```diff
@@ -0,0 +1,238 @@
+                trigger.Reason != AsyncCompletionData.CompletionTriggerReason.InvokeAndCommitIfUnique && 
+                !service.ShouldTriggerCompletion(sourceText, triggerLocation.Position, GetRoslynTrigger(trigger, triggerLocation)))
+            {
+                return AsyncCompletionData.CompletionStartData.DoesNotParticipateInCompletion;
+            }
+
+            return new AsyncCompletionData.CompletionStartData(
+                participation: AsyncCompletionData.CompletionParticipation.ProvidesItems,
+                applicableToSpan: new SnapshotSpan(triggerLocation.Snapshot, service.GetDefaultCompletionListSpan(sourceText, triggerLocation.Position).ToSpan()));
+        }
+
+        public async Task<AsyncCompletionData.CompletionContext> GetCompletionContextAsync(
+            IAsyncCompletionSession session, 
+            AsyncCompletionData.CompletionTrigger trigger, 
+            SnapshotPoint triggerLocation, 
+            SnapshotSpan applicableToSpan, 
+            CancellationToken cancellationToken)
+        {
+            var document = triggerLocation.Snapshot.GetOpenDocumentInCurrentContextWithChanges();
+            if (document == null)
+            {
+                return new AsyncCompletionData.CompletionContext(ImmutableArray<VSCompletionItem>.Empty);
+            }
+
+            var completionService = document.GetLanguageService<CompletionService>();
+
+            var completionList = await completionService.GetCompletionsAsync(
+                document,
+                triggerLocation,
+                GetRoslynTrigger(trigger, triggerLocation)).ConfigureAwait(false);
+
+            if (completionList == null)
+            {
+                return new AsyncCompletionData.CompletionContext(ImmutableArray<VSCompletionItem>.Empty);
+            }
+
+            var filterCache = new Dictionary<string, AsyncCompletionData.CompletionFilter>();
+
+            var itemsBuilder = new ArrayBuilder<VSCompletionItem>();
+            foreach (var roslynItem in completionList.Items)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+                var item = Convert(document, roslynItem, completionService, filterCache);
+
+                // Have to store the snapshot to reuse it in some projections related scenarios
+                // where data and session in further calls are able to provide other snapshots.
+                item.Properties.AddProperty(TriggerSnapshot, triggerLocation.Snapshot);
+                itemsBuilder.Add(item);
+            }
+
+            var items = itemsBuilder.ToImmutableAndFree();
+
+            var suggestionItemOptions = completionList.SuggestionModeItem != null
+                    ? new AsyncCompletionData.SuggestionItemOptions(
+                        completionList.SuggestionModeItem.DisplayText,
+                        (completionList.SuggestionModeItem.Properties.TryGetValue(Description, out var description)
```

Comment:
can't depend on looking at properties of a completion item.  Any service/provider may have stored this data however they wanted.  have to use service.GetDescriptionAsync if you truly need this data here.